src/metaopt/algo/space.py

```diff
@@ -0,0 +1,539 @@
+        args : list
+        kwargs : dict
+           See Parameters of `Dimension.__init__`.
+
+        """
+        kwargs['discrete'] = True
+        super(Integer, self).__init__(name, prior, *args, **kwargs)
+
+    def __contains__(self, point):
+        """Check if constraints hold for this `point` of `Dimension`.
+
+        :param x: a parameter corresponding to this `Dimension`.
+        :type x: numeric or array-like
+
+        `Integer` will check whether `point` contains only integers.
+
+        """
+        point_ = numpy.asarray(point)
+        if not numpy.all(numpy.equal(numpy.mod(point_, 1), 0)):
+            return False
+        return super(Integer, self).__contains__(point)
+
+
+class Categorical(Dimension):
+    """Subclass of `Dimension` for representing integer parameters.
+
+    Attributes
+    ----------
+    name : str
+    type : str
+    prior : `scipy.stats.distributions.rv_generic`
+    shape : tuple
+       See Attributes of `Dimension`.
+    discrete : bool
+       True
+    categories : tuple
+       A set of unordered stuff to pick out from, except if enum
+
+    """
+
+    def __init__(self, name, categories, **kwargs):
+        """Search space dimension that can take on categorical values.
+
+        Parameters
+        ----------
+        name : str
+           See Parameters of `Dimension.__init__`.
+        categories : dict or other iterable
+           A dictionary would associate categories to probabilities, else
+           it assumes to be drawn uniformly from the iterable.
+        kwargs : dict
+           See Parameters of `Dimension.__init__` for general.
+
+        """
+        kwargs['discrete'] = True
+        if isinstance(categories, dict):
+            self.categories = tuple(categories.keys())
+            self._probs = tuple(categories.values())
+            if sum(self._probs) != 1:
+                raise ValueError("Probabilities given for categories do not sum to one.")
+        else:
+            self.categories = tuple(categories)
+            self._probs = tuple(numpy.tile(1. / len(categories), len(categories)))
+
+        self._inverse = numpy.vectorize(lambda x: self.categories[x],
+                                        otypes=[numpy.object])
+        self._check = numpy.vectorize(lambda x: x in self.categories)
+
+        prior = distributions.rv_discrete(values=(list(range(len(self.categories))),
+                                                  self._probs))
+        super(Categorical, self).__init__(name, prior, **kwargs)
+
+    def sample(self, n_samples=1, seed=None):
+        """Draw random samples from `prior`.
+
+        .. seealso:: `Dimension.sample`
+
+        """
+        positions = super(Categorical, self).sample(n_samples, seed)
+        samples = list(map(self._inverse, positions))
+        result = []
+        for i, sample in zip(positions, samples):
+            if sample.shape == ():
+                result.append(self.categories[i].__class__(sample))
```

Comment:
The goal is to recast from `numpy.object` to what it was. I don't remember the case exactly.

Reply:
I don't understand. The algorithm work on the index space, and then we inverse it back to the categories. Why would the categories be numpy.objects?

Reply:
I use scipy's `rv_discrete` constructor, which accepts a tuple of (list of numbers, list of probabilities) - it does not accept a list of objects for some reason. But it was for coherency and for letting me assign probabilities to discrete items that I chose it.

Reply:
What's the advantage over sampling indexes using the given probabilities? We would not need _inverse and _check.

Reply:
That's what is being done here, sampling over indeces using `rv_discrete`. Try to execute small examples using the code itself with iPython to check what are the return types of the stuff each thing returns and why something is necessary or it isn't.

Reply:
I don't see the advantage of using rv_discrete and probably won't see it either after executing the code. The way I see it, rv_discrete is used so that Integer and Categorical can use the same method _Discrete.sample. In other words, we add complexity to Categorical so that it can share an almost empty method with Integer.

Reply:
I found out that scikit-optimizer is using a similar design for their Categorical dimension, using rv_discrete (see [here](https://github.com/scikit-optimize/scikit-optimize/blob/master/skopt/space/space.py#L434-L436)). Still, I don't understand what is the advantage of using it. I see this as a bad decision to increase design coherence between Integer and Categorical for no gain while it obfuscates Categorical's code.